crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs

```diff
@@ -0,0 +1,112 @@
+use ruff_python_ast::{Ranged, Stmt};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_source_file::Locator;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for banned imports at module level. The banned imports are allowed inline, such as
+/// within a function definition or an `if TYPE_CHECKING:` block.
+///
+/// ## Why is this bad?
+/// Some modules take a long time to import. Library authors might want to ensure that you only pay
+/// the import cost for these modules if you directly use them, rather than if you import a module
+/// that happens to use an expensive module in one of its functions.
+///
+/// ## Options
+/// - `flake8-tidy-imports.banned-module-level-imports`
+#[violation]
+pub struct BannedModuleLevelImports {
+    name: String,
+}
+
+impl Violation for BannedModuleLevelImports {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let BannedModuleLevelImports { name } = self;
+        format!("`{name}` is banned at module level. Please move the import inline.")
+    }
+}
+
+enum NameMatchPolicy {
+    ExactOnly,
+    ExactOrParents,
+}
+
+fn banned_at_module_level_with_policy<T>(
+    checker: &mut Checker,
+    name: &str,
+    stmt: &Stmt,
+    located: &T,
```

Comment:
nit: you can avoid the generic and pass the `TextRange` directly instead

Reply:
Good idea, thanks!